File: testers/avro/java-spring/generated-and-checked-in/com/example/events/OrderEventsPublisher.java
package com.example.events;

import java.util.concurrent.CompletableFuture;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.kafka.support.SendResult;
import org.springframework.stereotype.Service;

/** Type-safe event publisher for order-events topic */
@Service
public record OrderEventsPublisher(KafkaTemplate<String, OrderEvents> kafkaTemplate, String topic) {
  public OrderEventsPublisher(KafkaTemplate<String, OrderEvents> kafkaTemplate) {
    this(kafkaTemplate, "order-events");
  }

  public OrderEventsPublisher withKafkaTemplate(KafkaTemplate<String, OrderEvents> kafkaTemplate) {
    return new OrderEventsPublisher(kafkaTemplate, topic);
  }

  public OrderEventsPublisher withTopic(String topic) {
    return new OrderEventsPublisher(kafkaTemplate, topic);
  }

  /** Publish a OrderCancelled event */
  public CompletableFuture<SendResult<String, OrderEvents>> publish(
      String key, OrderCancelled event) {
    return kafkaTemplate.send(topic, key, event);
  }

  /** Publish a OrderPlaced event */
  public CompletableFuture<SendResult<String, OrderEvents>> publish(String key, OrderPlaced event) {
    return kafkaTemplate.send(topic, key, event);
  }

  /** Publish a OrderUpdated event */
  public CompletableFuture<SendResult<String, OrderEvents>> publish(
      String key, OrderUpdated event) {
    return kafkaTemplate.send(topic, key, event);
  }
}

File: testers/avro/scala-cats/generated-and-checked-in/com/example/events/OrderEventsPublisher.scala
package com.example.events

import cats.effect.IO
import fs2.kafka.KafkaProducer
import fs2.kafka.ProducerRecord
import fs2.kafka.ProducerRecords
import fs2.kafka.ProducerResult
import scala.annotation.unused

@unused
/** Type-safe event publisher for order-events topic */
case class OrderEventsPublisher(
  kafkaTemplate: KafkaProducer[IO, String, OrderEvents],
  topic: String = "order-events"
) {
  /** Publish a OrderCancelled event */
  def publish(
    key: String,
    event: OrderCancelled
  ): IO[ProducerResult[String, OrderEvents]] = {
    return kafkaTemplate.produce(ProducerRecords.one(ProducerRecord(topic, key, event))).flatten
  }

  /** Publish a OrderPlaced event */
  def publish(
    key: String,
    event: OrderPlaced
  ): IO[ProducerResult[String, OrderEvents]] = {
    return kafkaTemplate.produce(ProducerRecords.one(ProducerRecord(topic, key, event))).flatten
  }

  /** Publish a OrderUpdated event */
  def publish(
    key: String,
    event: OrderUpdated
  ): IO[ProducerResult[String, OrderEvents]] = {
    return kafkaTemplate.produce(ProducerRecords.one(ProducerRecord(topic, key, event))).flatten
  }
}